Fix #3521: changing the expected presentation connection state into "connecting"

tomoyukilabs opened the pull request:
This PR fixes #3521. Both controlling-ua/startNewPresentation_success-manual.html and controlling-ua/defaultRequest_success-manual.html are modified so that they examine if the initial state of presentation connection is `connecting` or not.

This PR also contains the following additional modifications:
- A test for `PresentationRequest.onconnectionavailable` is added.
- `defaultRequest` test will start via a user gesture on browser's user interface (i.e. "Cast..." menu in the case of Google Chrome).

tomoyukilabs committed:
Change the initial connection state into "connecting"

tomoyukilabs committed:
minor revision

tidoust commented on presentation-api/controlling-ua/PresentationRequest_onconnectionavailable-manual.html:
```diff
@@ -0,0 +1,62 @@
+                var connection = evt.connection;
+
+                // Check the initial state of a presentation connection
+                test(function() {
```
I haven't checked, but I'm not sure it's good practice to nest calls to `async_test` and `test`, especially because it means the number of test outcomes that appear in the final result depends on the context. For instance, here, if a timeout occurs before the `connectionavailable` event is fired, only the async test line will appear as `FAIL`. The other tests will simply not show up.

I would drop the `test` lines here and rather complete the `assert_xxx` statements with the failure reason as in:
`assert_equals(connection.state, 'connecting', 'The initial state of the presentation connection should be "connecting".');`

tidoust commented on presentation-api/controlling-ua/defaultRequest_success-manual.html:
```diff
@@ -2,12 +2,13 @@
-<p>Click the button below and select the available casting device, to start the manual test.</p>
-<button onclick="startPresentation()">Start Presentation Test</button>
+<p>Click the button or the menu item for presentation on your browser (for example, "Cast"), to start the manual test.</p>
+<p>If your browser does not support <code>defaultRequest</code>, this test will not begin.</p>
```
The test outcome should get reported to the test runner no matter what. There is no way to represent an optional test for the time being, I think. Last time I checked, the preferred outcome was to have the test FAIL. We could probably also prefix the test's label with `[Optional]` so that it's easy to spot that the failure may simply indicate lack of support for the feature.

To have the test fail, what should the user do? I suspect that if she just closes the window, she will have to state whether the test succeeded or failed to the test runner. It may be better to add a button that makes the test fail and instruct the user to press that button if the browser does not support `defaultRequest`.

That's open for discussion. I'm not saying this is the right approach. It's certainly not perfect :)

tidoust commented on presentation-api/controlling-ua/defaultRequest_success-manual.html:
```diff
@@ -22,29 +23,31 @@
+        navigator.presentation.defaultRequest.onconnectionavailable = t.step_func_done(function (evt) {
+            var connection = evt.connection;
+
+            test(function() {
```
Same comment on nested calls to test as above.

tidoust commented on presentation-api/controlling-ua/startNewPresentation_success-manual.html:
```diff
@@ -31,8 +32,10 @@
-                        // assert case for the promise_test
-                        assert_equals(connection.state, "connected", "The presentation has an connected state.");
+                        // Check the initial state of a presentation connection
+                        test(function() {
```
Similarly, I would rewrite this test with an enclosing `async_test` call and no `test` calls to ensure consistency between a PASS and a FAIL outcome.

tidoust reviewed (commented):
See suggested changes. I realize that some of them (nesting of test calls typically) applied to the initial tests as well, we could address them in a later PR if needed.

tomoyukilabs committed:
Removed nested test()
Added '_Optional_' prefix to defaultRequest test
Added 'not supported' button to defaultRequest test

tomoyukilabs committed:
Cancel changing the file name

tomoyukilabs commented:
@tidoust Thanks for your review!

> I would drop the `test` lines here and rather complete the `assert_xxx` statements with the failure reason

Okay, I removed nested `test()` and moved failure reason to `assert_xxx`.

> We could probably also prefix the test's label with `[Optional]` so that it's easy to spot that the failure may simply indicate lack of support for the feature.

I added `[Optional]` prefix to the test name.

> It may be better to add a button that makes the test fail and instruct the user to press that button if the browser does not support `defaultRequest`.

For now, I added the button to make the test fail. Since the current spec does not provide any hint about availability of `defaultRequest` for web apps, I cannot imagine any solution but using such a button...

tidoust commented:
Thanks. That all looks good to me. I'll let @louaybassbouss take a final look and merge this.

tidoust commented on presentation-api/controlling-ua/PresentationRequest_onconnectionavailable-manual.html:
```diff
@@ -0,0 +1,54 @@
+
+
+<script>
+    // ----------
```
Since this is a manual test, there needs to be a call somewhere to:
`setup({explicit_timeout: true});`

The test harness will time out after a few seconds otherwise. Note that one could argue that the test harness should automatically disable the timeout for tests that have "-manual" in their name. That's note the case today, though, so we need to do it explicitly in the code.

(I note I missed that in previous tests too... We'll fix other tests later on, no big deal)

tidoust commented on presentation-api/controlling-ua/PresentationRequest_onconnectionavailable-manual.html:
```diff
@@ -0,0 +1,54 @@
+                assert_true(connection instanceof PresentationConnection, 'The connection is an instance of PresentationConnection.');
+            });
+
+            request.start();
```
Another problem here is that there's no code to handle the case where the call to `start` fails. The Promise will be rejected but that will go unnoticed. To fix this, `promise_test` should be used instead of `async_test`, and the Promise should be returned as in:

```js
promise_test(function (t) {
  request.onconnectionavailable = [...];
  return request.start();
});
```

tidoust reviewed (commented):
@tomoyukilabs Reviewing the tests for other reasons (*), I realize that I missed a couple of things in my previous review. See inline.

(*) I'm preparing the migration of presentation URLs to a shared config file to avoid hardcoding the Cast app ID in each and every file.

tomoyukilabs committed:
added explicit_timeout

fixed PresentationRequest.onconnectionavailable test:
* added codes to handle the case where request.start() fails
* added codes to handle the case where the connectionavailable event is not fired although the presentation is started

tomoyukilabs committed:
minor revision

tomoyukilabs commented:
Thanks, @tidoust

> Since this is a manual test, there needs to be a call somewhere to:
`setup({explicit_timeout: true});`

Sorry, I forgot adding this code. I added this to the tests.

> Another problem here is that there's no code to handle the case where the call to `start` fails.

I changed `async_test` into `promise_test`, and added codes to handle the two cases:
* the case where the call to `start` fails
* the case where `connectionavailable` event is not fired even though the call to `start` does not fail

tidoust commented on presentation-api/controlling-ua/PresentationRequest_onconnectionavailable-manual.html:
```diff
@@ -0,0 +1,67 @@
+                request.start().then(function() {
+                    // This test fails if request.onconnectionavailable is not invoked although the presentation is started successfully
+                    t.step_timeout(function() { reject('The connectionavailable was not fired.') }, 5000);
+                }, function() {
```
I may be nitpicking here, but wouldn't it be better to replace `function() { reject(...); }` by "reject"?

```js
request.start().then(function() { ... }, reject);
```

This would expose the actual reason for failure in the test result in particular (`NotAllowedError`, `NotFoundError`, etc.) which would help with debugging.

tidoust reviewed (commented):
Thanks, @tomoyukilabs 

> the case where connectionavailable event is not fired although the call to start does not fail

Good catch. I had not thought about that either :)

One last comment on my side.

tomoyukilabs committed:
make the codes to handle `reject` more appropriate

tidoust commented:
OK, I think we've done enough back and forth on this one ;)
@louaybassbouss I'll merge it now so that I can work on the latest version of the test suite for the Cast URL update. Let's use another PR if further fixes are needed.

tomoyukilabs commented:
Thanks for reviewing, @tidoust!
